### PR TITLE
ensure etc-host file permission is 644 whatever umask is

### LINKED ISSUE #96342: Pod cannot work due to failing to access /etc/hosts inside container if umask of kubelet is configured as 027
**What happened**:
Process running as non-root will fail to work as it cannot access /etc/hosts inside the container when umask of kubelet  is configured as 027.

It is because the ioutil.WriteFile() used to create /etc/hosts is impacted by umask. Even though kubelet specify the permission of 0644 explicitly, with umask of 0027, it will be configured as 0640. As a result, application running inside the pods with non-root user, failed to access it.
**What you expected to happen**:
kubelet should pre-configure required umask by itself, instead of depending on the environment. At least we should document the required umask.

**How to reproduce it (as minimally and precisely as possible)**:
Lauch kubelet with umask of 027

**Anything else we need to know?**:
It's the same problem  https://github.com/kubernetes/kubernetes/issues/80668 figured out.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.19
- Cloud provider or hardware configuration: N/A
- OS (e.g: `cat /etc/os-release`): Ubuntu 18.04.4 LTS
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by luozhiwenn**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

ensure etc-host file permission is 644 whatever umask of kubelet is

#### Which issue(s) this PR fixes:

Fixes  [#96342](https://github.com/kubernetes/kubernetes/issues/96342), [#80668](https://github.com/kubernetes/kubernetes/issues/80668)

#### Special notes for your reviewer:

When the umask of kubelet is not `000`, the permission on the etc-hosts files created by kubelet is unexpected, which causes non-root applications to be unable to access these files.
Explicitly modify the files permissions to ensure that the files permissions meet the expected `644`.

#### Does this PR introduce a user-facing change?

```release-note
kubelet will ensure /etc/hosts file is mode 0644 regardless of umask.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
Is there a reason to add this specific chmod vs say, altering umask in the process? What's the use case where we'd want to have a custom umask only for some of kubelet's files?

**Comment by luozhiwenn** [comment]
> Is there a reason to add this specific chmod vs say, altering umask in the process? What's the use case where we'd want to have a custom umask only for some of kubelet's files?

Hi, is the following change what you refer to  `altering umask in the process`?
```go
func ensureHostsFile(fileName string, hostIPs []string, hostName, hostDomainName string, hostAliases []v1.HostAlias, useHostNetwork bool) error {
        // ...
        oldMask, _ := util.Umask(0)
	err = ioutil.WriteFile(fileName, hostsFileContent, 0644)
	util.Umask(oldMask)
        return err
}
```

There is a concurrency problem in `altering umask in the process` because method `ensureHostsFile` is invoked concurrently:

1. kubelet with umask `027`
2. Creating Pod A etc-hosts file, save oldMask `027` and set umask to `000`
3. Concurrently creating Pod B etc-hosts file, save oldMask `000` and set umask to `000`
4. Creates Pod A etc-hosts file complete, set umask back to oldMask `027`
5. Creates Pod B etc-hosts file complete, set umask back to oldMask `000`
6. kubelet with umask `000`

This concurrency problem causes the umask of the kubelet to be modified, therefore, I prefer to explicitly correct permissions with chmod rather than adding lock in this process.

Inspired by https://github.com/kubernetes/kubernetes/blob/76fe4039e5a50ceccce94f5c7bd94b6152ee9086/pkg/kubelet/kuberuntime/kuberuntime_container.go#L413

**Comment by BenTheElder** [comment]
No I am referring to syscall.Umask early in the process https://cs.opensource.google/go/go/+/go1.19.2:src/syscall/zsyscall_linux_amd64.go;l=874

**Comment by BenTheElder** [comment]
Or alternately: 

What is the use case to not permanently alter the umask of kubelet early in the process?

What use case requires a umask for kubelet that does not require it here?

If we're going to intentionally ignore umask for this file, why not for the whole kubelet process?

**Comment by luozhiwenn** [comment]
Apologies for the delayed response.

> Or alternately:
> 
> What is the use case to not permanently alter the umask of kubelet early in the process?
> 
> What use case requires a umask for kubelet that does not require it here?
> 
> If we're going to intentionally ignore umask for this file, why not for the whole kubelet process?

- For security purposes, umask is configured by default(022 or 027?) in many Linux systems. We hope that the kubelet conforms to this configuration, so that files with excessive permissions are not created inadvertently. In particular, kubelet is usually run as the root user, and the owner of the files created by kubelet is also root, which shuld be careful.
- Different from other files created by kubelet (apps are not interested in), applications read the /etc/hosts file in a typical scenario. Therefore, we want to ensure that the permission on the file is readable by other users regardless of the umask of kubelet.

**Comment by bart0sh** [comment]
/priority important-longterm
/triage accepted
@assign @BenTheElder

**Comment by matthyx** [comment]
@luozhiwenn as I understand it, the bug occurs only when there is no `/etc/hosts` file, which forces `os.WriteFile` to create it with 0644 + umask?
I hope there isn't a possible race condition between the time you create the file, and when you set the proper permissions?

**Comment by BenTheElder** [comment]
Forgot to comment: I unassigned because I'm not an approver here and someone who is needs to decide on the approach.

**Comment by BenTheElder** [comment]
> I hope there isn't a possible race condition between the time you create the file, and when you set the proper permissions?

I don't think that's an issue because we're relaxing permissions, not tightening them.

This is unavoidable I think, unless maybe we write it to a temp location, chmod, then move it.

**Comment by BenTheElder** [comment]
/remove-lifecycle stale
@kubernetes/sig-node-pr-reviews

**Review by mrunalp — APPROVED** [review]

**Comment by BenTheElder** [comment]
I think maybe this warrants a release note? It's a behavior change.

**Comment by BenTheElder** [comment]
release note: "kubelet will ensure `/etc/hosts` file is mode 0644 regardless of umask" or something along those lines?

**Comment by BenTheElder** [comment]
In case we're updating the release notes, which you can do by editing the PR body:
/hold

Comment `/hold cancel` to remove the hold and merge when your ready (including if we're not changing release note and going ahead and merging as-is).

**Comment by luozhiwenn** [comment]
> In case we're updating the release notes, which you can do by editing the PR body: /hold
> 
> Comment `/hold cancel` to remove the hold and merge when your ready (including if we're not changing release note and going ahead and merging as-is).

done

/hold cancel

**Pull request merged**
